### PR TITLE
Story/3492

### PR DESCRIPTION
**Pull request by kdwyer**

### CODE REVIEW

File: addons/udes_stock/views/stock_picking_batch.xml [hunk 1]
```diff
@@ -20,6 +20,9 @@
         <xpath expr="//button[@name='cancel_picking']" position="replace">
           <button name="cancel_picking" string="Cancel Batch" confirm="Cancelling the batch will cancel all its pickings. Are you sure you wish to cancel this Batch Picking?" type="object" states="ready,waiting,in_progress"/>
         </xpath>
+        <xpath expr="//button[@name='cancel_picking']" position="after">
+          <button name="mark_as_todo" string="Mark as Todo" confirm="Are you sure you want to mark this batch as todo?" type="object" states="draft"/>
```
Comment (ws-mps) [inline addons/udes_stock/views/stock_picking_batch.xml]: Only visible when state = draft.

Comment (kdwyer) [reply addons/udes_stock/views/stock_picking_batch.xml]: I thought the purpose was to move from draft to waiting?

File: addons/udes_stock/models/stock_picking_type.py [hunk 2]
```diff
@@ -406,3 +433,100 @@ def get_picking_type(self, picking_type_id):
...
+                ('picking_type_id', '=', picking_type.id),
+                ('state', '=', 'confirmed')]
+            limit = 1
+
+            processed = StockPickings.browse()
+            while to_reserve > 0 or reserve_all:
+                domain = base_domain[:]
+                if processed:
+                    domain.append(('id', 'not in', processed.ids))
+                stock_pickings = pickings or StockPickings.search(domain,
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_picking_type.py]: Filter pickings by picking type

File: addons/udes_stock/models/stock_picking_type.py [hunk 3]
```diff
@@ -406,3 +433,100 @@ def get_picking_type(self, picking_type_id):
...
+        """
+        Reserve stock according to the number of reservable pickings.
+
+        0 reservable pickings means this function should not reserve stock
+        -1 reservable picking means all reservable stock should be reserved.
+
+        Arguments:
+        picking -- optional; a single picking to be reserved.
+        """
+        StockPickings = self.env['stock.picking']
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_picking_type.py]: For consitency, Picking or StockPicking

File: addons/udes_stock/models/stock_picking_type.py [hunk 4]
```diff
@@ -406,3 +433,100 @@ def get_picking_type(self, picking_type_id):
...
+                processed |= stock_pickings
+
+                if not picking_type.u_handle_partials:
+                    if not all(x == 'assigned' for x in
+                               stock_pickings.mapped('state')):
+                        if pickings:
+                            # we need to construct our error message before the
+                            # changes are rolled back.
+                            key = lambda x: x.state != 'assigned'
+                            unsatisfied = stock_pickings.filtered(key)
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_picking_type.py]: You can compute unsatisfied before line 511, and change condition of current line 511 to `if unsatisfied:`

File: addons/udes_stock/wizard/reservation_views.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        PickingType = self.env['stock.picking.type']
+        self.ensure_one()
+        picking_ids = self.env.context.get('active_ids')
+
+        _logger.info("User %r has requested stock reservation for pickings %r",
+                     self.env.uid, picking_ids)
+
+        pickings = Picking.browse(picking_ids)
+        for _key, group in pickings.groupby(lambda x: x.batch_id):
+            reserve_batches = (group.mapped('picking_type_id')
```
Comment (ws-mps) [inline addons/udes_stock/wizard/reservation_views.py]: It is allowed to use dot notation at mapped.

File: addons/udes_stock/models/stock_picking_type.py [hunk 6]
```diff
@@ -406,3 +433,104 @@ def get_picking_type(self, picking_type_id):
 
     def get_action_picking_tree_draft(self):
         return self._get_action('udes_stock.action_picking_tree_draft')
+
+    def reserve_stock(self, pickings=None):
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_picking_type.py]: I would move this function to stock.picking.

**Pull request merged**